### PR TITLE
feat: added nightly feature gate

### PR DESCRIPTION
**Pull request by Yandrik**
This should enable the nrf-softdevice crate to be used on stable rust.
Note that the only tests I performed are building it with and without the nightly feature flag. So please do double-check that I didn't blow anything up.
A non-async flash write implementation might also be a good idea. I have one, but it's not really production ready. Basically just the async one, but ripped apart without the async stuff. Do tell whether I should add that.

### CODE REVIEW

File: nrf-softdevice-macro/src/lib.rs [hunk 1]
```diff
@@ -48,12 +48,15 @@ pub fn gatt_server(_args: TokenStream, item: TokenStream) -> TokenStream {
...
         _ => {
-            struc
+            let s = struc
                 .ident
                 .span()
-                .unwrap()
-                .error("gatt_server structs must have named fields, not tuples.")
+                .unwrap();
+
+            #[cfg(feature = "nightly")]
```
Comment (Dirbaio) [inline nrf-softdevice-macro/src/lib.rs:53]: This means errors won't be reported at all without nightly, the macro will silently allow wrong things through.

Perhaps copy this, that allows having the same functionality on stable https://github.com/embassy-rs/embassy/blob/master/embassy-macros/src/util/ctxt.rs

example on how to use it: https://github.com/embassy-rs/embassy/blob/master/embassy-macros/src/macros/main.rs#L71-L84

Comment (Yandrik) [reply nrf-softdevice-macro/src/lib.rs:53]: Oh, i expected it to just blow up without any proper error reporting. On it!

**Comment by Dirbaio** [comment]
Could you add nightly+stable checking in CI? so we don't accidentally introduce nightly-only features when building for stable.
- in `ci.sh` make sure everything gets built with the `nightly` feature.
- copy to `ci_stable.sh`, make it not enable `nightly` and not build the examples, add a workflow for it.

File: nrf-softdevice-macro/src/lib.rs [hunk 2]
```diff
@@ -42,16 +44,26 @@ struct Characteristic {
 
 #[proc_macro_attribute]
 pub fn gatt_server(_args: TokenStream, item: TokenStream) -> TokenStream {
+
+    // Context for error reporting
+    #[cfg(not(feature = "nightly"))]
```
Comment (Dirbaio) [inline nrf-softdevice-macro/src/lib.rs:49]: We can unconditionally use the `Ctxt` for reporting errors, even on nightly. Makes the code simpler.

Comment (Yandrik) [reply nrf-softdevice-macro/src/lib.rs:49]: got it

**Comment by Yandrik** [comment]
> Could you add nightly+stable checking in CI? so we don't accidentally introduce nightly-only features when building for stable.
> 
>     * in `ci.sh` make sure everything gets built with the `nightly` feature.
> 
>     * copy to `ci_stable.sh`, make it not enable `nightly` and not build the examples, add a workflow for it.

The examples auto-depend on the nightly feature now, so they'll build with that by default.
I'll remove them from the stable ci file

**Comment by Yandrik** [comment]
Should be done & working, tho I haven't actually checked macro error reporting

**Comment by Yandrik** [comment]
I believe that the macro checks aren't quite right yet, tho they do panic at least

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -25,5 +25,7 @@ jobs:
 
       - name: Check fmt
         run: cargo fmt -- --check
-      - name: Build
+      - name: Build (nightly)
         run: ./ci.sh
+      - name: Build (stable)
+        run: ./ci-stable.sh
```
Comment (Dirbaio) [inline .github/workflows/rust.yml:31]: could you make this run with latest Rust stable? Currently it'll be picking up the nightly version from `rust-toolchain.toml`.

Comment (Yandrik) [reply .github/workflows/rust.yml:31]: done

File: nrf-softdevice/Cargo.toml [hunk 4]
```diff
@@ -32,7 +32,7 @@ ble-sec = []
 
 critical-section-impl = ["critical-section/restore-state-bool"]
 
-nightly = []
+nightly = ["embedded-storage-async"]
```
Comment (Dirbaio) [inline nrf-softdevice/Cargo.toml:35]:
```suggestion
nightly = ["dep:embedded-storage-async"]
```

optional deps implicitly create a Cargo feature, unless you mention them with `dep:`. This'll disallow people from doing `nrf-softdevice = { .., features = ["embedded-storage-async"] }`, which would compile but not work.

Comment (Yandrik) [reply nrf-softdevice/Cargo.toml]: done

**Comment by Yandrik** [comment]
I don't quite get why GitHub Actions has a problem with building futures-core here, it works on my machine. Could you take a look?

**Comment by Dirbaio** [comment]
It needs `rustup target add`. With nightly it's done by the `rust-toolchain.toml` file, but with `+stable` it dosen't. 

Perhaps the easiest fix is to a `rust-toolchain-stable.toml` file, and `mv` it to `rust-toolchain.toml` when doing stable testing..

**Comment by Dirbaio** [comment]
Looks great! Thank you for your work :)

**Pull request merged**